### PR TITLE
Cast platform agnostically

### PR DESCRIPTION
**Pull request by adityagoel4512**
The signedness of char (the [desired](https://github.com/microsoft/onnxruntime/blob/6baaaf516538f9059da3558b2cd22128a9e42c07/onnxruntime/core/framework/error_code.cc#L12) type) is undefined and is platform dependent. It's probably worth setting up the infrastructure to verify this is the only instance by broadening the test matrix.

I haven't used nix before so some help with the Windows failure would be great.

### CODE REVIEW

File: .github/workflows/nix.yml [hunk 1]
```diff
@@ -7,8 +7,12 @@ env:
...
 jobs:
   build:
-
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os:
+          - ubuntu-latest
+          - macos-latest
```
Comment (SaravananSathyanandhaQC) [inline .github/workflows/nix.yml:15]: Might want to add `ubuntu-latest-arm-2core` if you want to test on there too (not sure if you need to opt into the beta arm runners somewhere first)

Comment (adityagoel4512) [reply .github/workflows/nix.yml:15]: The link to the private beta seems to be [invalid/expired](https://github.blog/changelog/2023-10-30-accelerate-your-ci-cd-with-arm-based-hosted-runners-in-github-actions/) :(

**Comment by cbourjau** [comment]
I'll migrate the repository from nix to conda

**Comment by adityagoel4512** [comment]
It would be nice to also do something with qemu considering there aren't public Linux ARM runners.

**Comment by adityagoel4512** [comment]
[This](https://docs.docker.com/build/ci/github-actions/multi-platform/) looks promising for AArch64 testing.

**Review by cbourjau — APPROVED** [review]

**Comment by adityagoel4512** [comment]
Migrating to conda later then?

**Comment by cbourjau** [comment]
I'll merge this for now. The cross-compilation to `linux_aarch64` seems to work, but I'm getting linker error (due to the cross-compiling). This is certainly better than what we have now.

**Comment by adityagoel4512** [comment]
Sure - I think we may as well go for https://github.com/cross-rs/cross if it takes time to get cross-compiling working from scratch.

**Pull request merged**